internal/controller/pkg/revision/runtime.go

```diff
@@ -0,0 +1,253 @@
+	namespace                 string
+	serviceAccountPullSecrets []corev1.LocalObjectReference
+	controllerConfig          *v1alpha1.ControllerConfig
+}
+
+// NewRuntimeManifestBuilder returns a new RuntimeManifestBuilder.
+func NewRuntimeManifestBuilder(ctx context.Context, client client.Client, namespace string, serviceAccount string, pwr v1.PackageRevisionWithRuntime) (*RuntimeManifestBuilder, error) {
+	b := &RuntimeManifestBuilder{
+		namespace: namespace,
+		revision:  pwr,
+	}
+
+	if ccRef := pwr.GetControllerConfigRef(); ccRef != nil {
+		cc := &v1alpha1.ControllerConfig{}
+		if err := client.Get(ctx, types.NamespacedName{Name: ccRef.Name}, cc); err != nil {
+			return nil, errors.Wrap(err, errGetControllerConfig)
+		}
+		b.controllerConfig = cc
+	}
+
+	sa := &corev1.ServiceAccount{}
+	// Fetch XP ServiceAccount to get the ImagePullSecrets defined there.
+	// We will append them to the list of ImagePullSecrets for the runtime
+	// ServiceAccount.
+	if err := client.Get(ctx, types.NamespacedName{Namespace: namespace, Name: serviceAccount}, sa); err != nil {
+		return nil, errors.Wrap(err, errGetServiceAccount)
+	}
+	b.serviceAccountPullSecrets = sa.ImagePullSecrets
+
+	return b, nil
+}
+
+// ServiceAccount builds and returns the ServiceAccount manifest.
+func (b *RuntimeManifestBuilder) ServiceAccount(overrides ...ServiceAccountOverrides) *corev1.ServiceAccount {
+	sa := defaultServiceAccount(b.revision.GetName())
+
+	overrides = append(overrides,
+		// Currently it is not possible to override the namespace,
+		// ownerReferences or pullSecrets of the service account, and we could
+		// define them as defaults. However, we will leave them as overrides
+		// to indicate that we are opinionated about them currently and follow
+		// a consistent pattern.
+		ServiceAccountWithNamespace(b.namespace),
+		ServiceAccountWithOwnerReferences([]metav1.OwnerReference{meta.AsController(meta.TypedReferenceTo(b.revision, b.revision.GetObjectKind().GroupVersionKind()))}),
+		ServiceAccountWithAdditionalPullSecrets(append(b.revision.GetPackagePullSecrets(), b.serviceAccountPullSecrets...)),
+	)
+
+	if cc := b.controllerConfig; cc != nil {
+		overrides = append(overrides, ServiceAccountWithControllerConfig(cc))
+	}
+
+	for _, o := range overrides {
+		o(sa)
+	}
+
+	return sa
+}
+
+// Deployment builds and returns the Deployment manifest.
+func (b *RuntimeManifestBuilder) Deployment(serviceAccount string, overrides ...DeploymentOverrides) *appsv1.Deployment {
+	d := defaultDeployment(b.revision.GetName())
+
+	overrides = append(overrides,
+		DeploymentWithNamespace(b.namespace),
+		DeploymentWithOwnerReferences([]metav1.OwnerReference{meta.AsController(meta.TypedReferenceTo(b.revision, b.revision.GetObjectKind().GroupVersionKind()))}),
+		DeploymentWithSelectors(b.podSelectors()),
+		DeploymentWithServiceAccount(serviceAccount),
+		DeploymentWithImagePullSecrets(b.revision.GetPackagePullSecrets()),
+		DeploymentRuntimeWithImage(b.revision.GetSource()),
```

Comment:
we used to default to the revision's source, but overwrite it with the meta if specified, which is the cause of the usage e2e failure

Reply:
see https://github.com/crossplane/crossplane/pull/4767/files#diff-53e9c93d08b0ecb30843baebd9c68749a3bcc0640aa836972621e9a9243bb15fL125

Reply:
Good catch! Just pushed https://github.com/crossplane/crossplane/pull/4767/commits/bb11ae81672bf566f421b74de1cdd80e7f555104 to fix.